Port disjoint layout utils in rust

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit ports the disjoint layout utils into rust. The eventual goal of this is to enable porting the full DenseLayout and SabreLayout pass to rust so it's a standalone function. This will enable us exposing these passes as a C function. Since the eventual goal of this functionality is to expose it as a C function this attempts to do as much as possible without a py token. A standalone rust entrypoint is added which will serve as what is called from the standalone rust passes in the future, this is then wrapped in a python entrypoint.

One minor note is the visibility of the run pass over components entrypoint is `pub(crate)`, but the intent is for it to be `pub`. This visibility was limited here to avoid changing the target's visbility level (which is `pub(crate)` for some reason). This is already being fixed in #14056 and when that merges we can update this PR.

### Details and comments

~This is based on top of #14313 and will need to be rebased when that merges. You can see the contents of just this PR by looking at: https://github.com/Qiskit/qiskit/pull/14177/commits/4f6c986431f58ddfadb0ab7f214e758c4e00af00~ #14313 has merged and this is now rebased.

TODO:

- [x] Fix test failures

mtreinish commented:
This should be ready for review now. I'm not super happy about the split between python and rust, but I didn't think it was worth supporting the `CouplingMap` only workflow in Rust.

mtreinish commented:
I've marked this as on hold because the changes around `Vars` and `Stretches` in DAGCircuit are causing a merge conflict here. The main changes this made to the dag circuit were about py token isolation to avoid adding a `py` token to the functions being added here (the goal is to leverage this in the C API eventually). Since this all still in flux, while those are being moved to rust I'm just going to wait until #14176 merges and then I'll rebase this.

mtreinish commented:
This is mostly ready (besides needing to update the rust tests) but I'm leaving it on hold because I'm going to split out the py token changes to the dagcircuit. It's a logically separate change and it's a distraction from reviewing the disjoint layout changes. I'll open a new PR for the py token changes and then rebase this ontop of that new PR.

jakelishman commented on crates/accelerate/src/disjoint_layout.rs:
```diff
@@ -0,0 +1,488 @@
+static COUPLING_MAP: ImportOnceCell =
+    ImportOnceCell::new("qiskit.transpiler.coupling", "CouplingMap");
+
+type CouplingMap = StableDiGraph<NodeIndex, ()>;
```
Is there a need for this to be a `StableGraph`?  Feels like it could just be a `Graph` - the node indices already have a meaning in it, so you can't actually remove them.

jakelishman commented on crates/accelerate/src/disjoint_layout.rs:
```diff
@@ -0,0 +1,488 @@
+    mut run_func: F,
+) -> PyResult<Option<Vec<T>>>
+where
+    F: FnMut(&DAGCircuit, &CouplingMap) -> PyResult<T>,
```
The ownership model interacting with Python is a bit nasty here, but it's also not great in the pure-Rust path either.  If we're coming from Python, we always get a reference in, but that means that we always have to clone the object to pass it _back_ to Python.

From the Rust side, having `F` take `DAGCircuit` by reference, even when we _have_ the owned version whenever we have to split feels unfortunate.  Given that we have to have mutability of the DAG for `separate_dag` to work currently, it feels like a pure-Rust path, if we didn't have to care about Python at all, would probably want `run_pass_over_connected_components` to take ownership of the `DAGCircuit` so that `F` can always take `DAGCircuit` by ownership, and consequently would let passes have full power of memory management (and having ownership is particularly nice if you need to do a lot of extensions, because you can use `DAGCircuitBuilder`, etc, though I guess that particular one could swap to being lifetime bound on a mutable reference).  At the moment, it's Python stopping us from doing that, and it'll be C too, with our current architecture.

I think the questions about ownership are worth having in mind for design of a complete transpiler pipeline in Rust.

jakelishman replied:
The ownership model interacting with Python is a bit nasty here, but it's also not great in the pure-Rust path either.  If we're coming from Python, we always get a reference in, but that means that we always have to clone the object to pass it _back_ to Python (as the todo comment notes).

From the Rust side, having `F` take `DAGCircuit` by reference, even when we _have_ the owned version whenever we have to split feels unfortunate.  Given that we have to have mutability of the DAG for `separate_dag` to work currently, it feels like a pure-Rust path, if we didn't have to care about Python at all, would probably want `run_pass_over_connected_components` to take ownership of the `DAGCircuit` so that `F` can always take `DAGCircuit` by ownership, and consequently would let passes have full power of memory management (and having ownership is particularly nice if you need to do a lot of extensions, because you can use `DAGCircuitBuilder`, etc, though I guess that particular one could swap to being lifetime bound on a mutable reference).  At the moment, it's Python stopping us from doing that, and it'll be C too, with our current architecture.

I think the questions about ownership are worth having in mind for design of a complete transpiler pipeline in Rust.  I'm not sure if it warrants a rearchitecture of `disjoint_utils` in this PR.

jakelishman commented on qiskit/transpiler/passes/layout/dense_layout.py:
```diff
@@ -66,11 +67,24 @@ def run(self, dag):
+                dag,
+                self.coupling_map,
+                self._inner_run,
+            )
```
Is this kind of thing the only reason to keep all the Python stuff around (barring backwards compat deprecations etc)?  If so, would it be ok to get rid of the Python-space use, and pass a `Target.from_configuration(basis_gates=["cx", "u"], coupling_map=self.coupling_map)` to the Rust-space version if/when needed?  It's a bit less efficient, but that's the path we want to drive towards, I think?

jakelishman commented on crates/accelerate/src/disjoint_layout.rs:
```diff
@@ -0,0 +1,488 @@
+        }
+        if len_args > 2 {
+            return Err(MultiQEncountered::new_err(""));
+        }
```
This is copied over from VF2, but in this situation, is it really an error during the disjoint handling to have a multi-q gate?  Feels like it should be up to the layout pass to decide that.

jakelishman reviewed (commented):
Honestly, I'm struggling a bit to work out if doing this is actually going to be helpful in its current form for Sabre or not.  Rust-space Sabre doesn't use `CouplingMap`; it takes `Target`, so the disjoint utils applying their opinion that everything should be `CouplingMap` (and one particular representation of the coupling map that's quite tied to one particular implementation of VF2 utils that I'm also obsoleting) isn't actually going to help.  That's a fairly minor tweak, though.

As a larger change: my feeling right now is that the old `run_pass_over_connected_components` is trying to do too much at once.  I think what I really _need_ for Sabre is:
- a method that takes a `Target` and a `DAGCircuit` and gives me something like
  ```rust
  struct DisjointComponent {
    /// A component of the target.
    sub_target: Target,
    /// Mapping of fake `PhysicalQubit`s in the `sub_target` (which are consecutive)
    /// to the actual `PhysicalQubit`s in the full `Target`.
    physical: Vec<PhysicalQubit>,
    sub_dag: DAGCircuit,
    /// Same as `physical`, but for the qubits in the `DAGCircuit`.
    /// `virtual.len() <= physical.len()`.
    virtual: Vec<VirtualQubit>,
  }
  // `None` means "no splitting necessary".  Add whatever other `Result` needed too.
  fn distribute_components(dag: &DAGCircuit, target: &Target) -> Option<Vec<DisjointComponent>>;
- the existing `combine_barriers`, but perhaps somehow wrapped in an API that makes it harder to forget to call it (though I suppose we have the same problem in Python right now anyway, so maybe never mind).

I guess my two most major things are:
- I want to work `Target`-native (but it's not the end of the world to get a `CouplingMap` for now - I can just make a dummy `Target` from it and then turn that back into a `CouplingMap` lol)
- I don't really want to have to try and contort the pass to fit a `run_func` abstraction.  It's easier and more efficient to let me do that bit myself, because then I can play tricks with lifetime-bound objects that refer to the split inputs, etc etc, and they live long enough for me to keep references to them, without needing a clone in the case that splitting _wasn't_ required.

jakelishman commented:
If we ever envisaged situations where we needed to call `distribute_components` more than once in a single compile pipeline, I'd change its proposed signature so that the splitting of the `Target` could be done only once, and then different `DAGCircuit`s distributed over its components as a separate step, but at least in the ways we currently do disjoint handling, that's getting ahead of myself.

mtreinish committed:
Port disjoint layout utils in rust

This commit ports the disjoint layout utils into rust. The eventual goal
of this is to enable porting the full DenseLayout and SabreLayout pass
to rust so it's a standalone function. This will enable us exposing
these passes as a C function. Since the eventual goal of this
functionality is to expose it as a C function this attempts to do as
much as possible without a py token. A standalone rust entrypoint is
added which will serve as what is called from the standalone rust passes
in the future, this is then wrapped in a python entrypoint.

If we have a target that can't build a coupling map from Rust the
disjoint utils return None. This could be because the target could be
global and not have any connectivity constraints, or it could be caused
by having a FakeTarget that doesn't expose certain constraints to rust.
This commit fixes this by falling back to the coupling map python path
in these cases since we can get the python coupling map and use the
python version of the disjoint utils.

mtreinish committed:
Improve ergonomics of rust interface

mtreinish committed:
Run rustfmt

jakelishman commented on crates/circuit/src/bit.rs:
```diff
@@ -63,6 +63,10 @@ impl<R: Register + PartialEq> BitLocations<R> {
         }
         None
     }
+
+    pub fn index(&self) -> u32 {
+        self.index
+    }
```
Super minor: usually I expect an `index` method to return a `usize`, so it can be directly used "as an index" (c.f. `NodeIndex::index`, etc).  If that's not the intent here, then no worries.

mtreinish replied:
I can change it although at the end of the day I guess I'm trying to get the `Qubit` which is a `u32`. I needed a way to go from a `ShareableQubit` to a `Qubit` for copying an instruction to a new sub-dag. I can work with whatever type you think is best here.

jakelishman commented on crates/transpiler/src/passes/disjoint_layout.rs:
```diff
@@ -0,0 +1,545 @@
+                .getattr(intern!(py, "graph"))?
+                .set_item(node.index(), cmap.node_weight(node).unwrap().index())?;
+        }
+        // TODO: Remove the dag clone
```
I think this comment is at least misplaced now?

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/14177/commits/5fdcf92a63fbe5a64456c85de4a3ac9722a6dedd

jakelishman commented on crates/transpiler/src/passes/disjoint_layout.rs:
```diff
@@ -0,0 +1,545 @@
+    NoneNeeded,
+    TargetSubset(Vec<PhysicalQubit>),
+    Arbitrary(Vec<DisjointComponent>),
+}
```
Some internal docstring comments here might help for the future.

mtreinish replied:
Added docstrings in: https://github.com/Qiskit/qiskit/pull/14177/commits/39a6680dc83f42f0fe1f2102d7f574c35eec477f

jakelishman commented on crates/transpiler/src/passes/disjoint_layout.rs:
```diff
@@ -0,0 +1,545 @@
+    Arbitrary(Vec<DisjointComponent>),
+}
+
+type CouplingMap = StableDiGraph<NodeIndex, ()>;
```
1. does this need to be a `StableGraph` over a `Graph`?  Also, is this graph _definitely_ directed?  Given that we've erased the basis-gates information from the `Target`, I'd argue that it's meaningless to try and call it directed and we should make it undirected.
2. which graph is the `NodeIndex` a reference into?  I _think_ eventually they refer to `PhysicalQubit`s in the `Target`?  Maybe we can just make the weights `PhysicalQubit`?

mtreinish replied:
I think I can probably get away with a `Graph` there shouldn't be any removals, but also the canonical index is the node weight which represents the `PhysicalQubit` (which I'll update).  I also think we can move it to be undirected, since in current state of the PR we're only getting a physical qubit list and not passing around a graph object directly.

mtreinish replied:
I updated this in: https://github.com/Qiskit/qiskit/pull/14177/commits/cd7ad0a72033eaa5b9f3e3746fda77e02480e0ef there is one small nuance that `CouplingMap` in python is explicitly directed, but it's simple enough to workaround this. It only effects dense layout since sabre always works with undirected graphs.

jakelishman commented on crates/transpiler/src/passes/disjoint_layout.rs:
```diff
@@ -0,0 +1,545 @@
+}
+
+pub fn distribute_components(dag: &mut DAGCircuit, target: &Target) -> PyResult<DisjointSplit> {
+    let coupling_map: CouplingMap = match build_coupling_map(target) {
```
We have an ordering choice here: I wrote https://github.com/Qiskit/qiskit/blob/9cbb8915fc9d1ec425082daeea1acb195d53b4c2/crates/transpiler/src/target/mod.rs#L1385-L1421 for Sabre, which has a bunch of overlap with this, but is a bit more unified and centralised to `Target`, and is more explicit about the assumptions it makes.  Perhaps we could pull that in (or I can separate it off to merge first) and add a variant that takes a `node_fn: impl FnMut(PhysicalQubit) -> T` to produce node weights?

mtreinish replied:
If you don't mind generalizing it that sounds good to me.  This also comes up in vf2 where there is a similar function already, I didn't make a generic version because this PR was racing with the vf2 one (I originally wrote them around the same time). But if you have a generic version we can deduplicate it all.

jakelishman replied:
I have my own version of this for my VF2 branch, and I _might_ be able to generalise this `Target` version of it to cover the case, but I can't remember entirely.

jakelishman replied:
I think let's just merge as is from the current PR head, and any changes we make in the future can be done as part of subsequent PRs anyway.

jakelishman commented on crates/transpiler/src/passes/disjoint_layout.rs:
```diff
@@ -0,0 +1,545 @@
+        .filter(|(_, dag_indices)| !dag_indices.is_empty())
+        .map(|(cmap_index, dag_indices)| {
+            let mut out_dag = dag_components[*dag_indices.first().unwrap()].clone();
+            for dag_index in &dag_indices[1..] {
+                let dag = &dag_components[*dag_index];
+                for qubit in dag.qubits().objects() {
+                    out_dag.add_qubit_unchecked(qubit.clone())?;
+                }
+                for clbit in dag.clbits().objects() {
+                    out_dag.add_clbit_unchecked(clbit.clone())?;
+                }
+                for qreg in dag.qregs() {
+                    out_dag.add_qreg(qreg.clone())?;
+                }
+                for creg in dag.cregs() {
+                    out_dag.add_creg(creg.clone())?;
+                }
+                out_dag.compose(
+                    dag,
+                    Some(dag.qubits().objects()),
+                    Some(dag.clbits().objects()),
+                    false,
+                )?;
+            }
```
This was pre-existing so we can leave it, but I think the call to `compose` here would cause errors if there are `Var` or `Stretch` instances in `dag`.

mtreinish replied:
Should I change `inline_captures` to `true` or do I need to do more work here? It's unlikely for this to come up in practice but it would still be good to fix this.

jakelishman replied:
I think the fix would be changing each `dag` component to be created from an initial empty-like template that sets `vars_mode="captures"`, then we'd change this call to `compose` to have `inline_captures=true`, but let's just follow up at some later point rather than trying to change it on-the-fly.

jakelishman commented on crates/transpiler/src/passes/disjoint_layout.rs:
```diff
@@ -0,0 +1,545 @@
+            Ok((out_dag, subgraph))
+        })
+        .collect::<PyResult<Vec<_>>>()?;
+    if out_component_pairs.len() == 1 {
+        return Ok(DisjointSplit::TargetSubset(
+            out_component_pairs[0]
+                .1
+                .node_weights()
+                .map(|x| PhysicalQubit::new(x.index() as u32))
+                .collect(),
+        ));
+    }
```
For the future: I think we want to do this check before we even split the DAG.  I feel like the structure should be that we split the `Target` first, and then only split the DAG if we have to.

mtreinish replied:
Oh that's simple to add, I can do it tomorrow it's just add an additional check before L152 where we separate the dag to check if the number of active qubits in the full dag is less than the largest component.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/14177/commits/f5b0d2b9c9f9638d985f0465dbfb9b1a9e7d29ae

jakelishman commented on crates/transpiler/src/passes/disjoint_layout.rs:
```diff
@@ -0,0 +1,545 @@
+                )?;
+            }
+        }
+    }
```
One day I'll convince you to use `let ... = ... else { continue; };` instead of indenting the whole block twice haha.

mtreinish replied:
This is what happens when I learn a language too early. I don't have newer language constructs in my muscle memory :) Let else was added in 1.65 which is too new, I'll get used to using it at some point.

mtreinish replied:
https://github.com/Qiskit/qiskit/pull/14177/commits/18a38752a6bc547b38de71ab84e9d473c89740ea

jakelishman reviewed (commented):
I have minor questions about the Rust interface, but it's not like we can't change it in the future, so mostly it's not a big deal.  There's a couple of comments left over from a previous review, but I don't think they're super important.

The most major question: can we not remove some of the Python-space versions of these functions?  I'm surprised that `disjoint_utils.py` isn't even in the diff.

mtreinish committed:
Remove stale comment

mtreinish committed:
Use Graph instead of StableGraph and UnGraph for coupling

This commit moves to using a regular Graph instead of StableGraph
internally in the disjoint utils. We are never removing nodes from
interaction graphs or coupling graphs so we don't need index stability
on removals and using Graph avoids the overhead for tracking the free
list. Additionally we can use undirected graphs for the coupling graph
since we don't have physical gates defined in the layout stage because
we need to pick the qubits first before we can worry about the
directionality of the gates selected. The one caveat with this is the
CouplingMap in Python is directed so we need to add the reverse edge
when making the coupling map object in Python to ensure we're not
representing things differently.

mtreinish committed:
Remove python implementation of disjoint utils

This commit removes the python space implementation of the disjoint
utils. This was originally kept around for targets without a coupling
map since the rust version is Target only. But we can handle this by
making a dummy target that contains minimal gates (u and cx) and
the coupling constraints. Adding this means we are no longer using the
python version of these functions and as they're internal utilities we
don't need to keep them anymore.

As a side effect one test using the disjoint handling via just a
coupling map is changed in this commit. This is a side effect of
internally using a undirected graph instead of a directed graph the
split components are the same but the sorted order is now turning up the
high qubits first instead of the lower qubits which results in placing
the circuit on the high qubit component (3, 4, 5) instead of the equally
sized component on the low qubits (0, 1, 2).

mtreinish committed:
Run rustfmt

mtreinish commented:
> The most major question: can we not remove some of the Python-space versions of these functions? I'm surprised that disjoint_utils.py isn't even in the diff.

Removed the python space stuff in: https://github.com/Qiskit/qiskit/pull/14177/commits/151ee6660108fbd5302fa4eaa4e30ed97aae8976 I left it around for the no-target case which only comes up in testing really. But I just used a dummy target to get around that.

mtreinish committed:
Add docs for distribute component return types

mtreinish committed:
Merge branch 'main' into port-disjoint-utils-rs

mtreinish committed:
Add up-front check on target subset result

mtreinish committed:
Let the wookie win

jakelishman reviewed (approved):
![giphy](https://github.com/user-attachments/assets/470483e8-23bc-4a3e-a384-ff06236809cc)
